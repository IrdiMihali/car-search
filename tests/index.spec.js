import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import 'jest-styled-components';

import Home from '@/pages/index';
import { responseMock } from './responseMock';

jest.useFakeTimers();

describe('Given `Home`', () => {
	beforeEach(() => {
		fetch.resetMocks();
	});

	it('should render correctly', () => {
		const { container } = render(<Home />);

		expect(container).toMatchSnapshot();
	});

	describe('When the user starts typing on the Pick-up Location input', () => {
		describe('and user enters 1 character', () => {
			it('should NOT make a fetch request', () => {
				render(<Home />);
				const input = screen.getByTestId('search-input');

				fireEvent.change(input, { target: { value: 'i' } });

				act(() => {
					jest.advanceTimersByTime(500);
				});

				expect(fetch).not.toHaveBeenCalled();
			});
		});

		describe('and the user enters more than 1 character', () => {
			let input;

			beforeEach(async () => {
				fetch.mockResponse(JSON.stringify(responseMock));

				render(<Home />);
				input = screen.getByTestId('search-input');

				fireEvent.change(input, { target: { value: 'Portugal' } });

				act(() => {
					jest.advanceTimersByTime(500);
				});

				await waitFor(() =>
					expect(screen.queryByTestId('loading-spinner')).toBeInTheDocument()
				);
				await waitFor(() =>
					expect(
						screen.queryByTestId('loading-spinner')
					).not.toBeInTheDocument()
				);
			});

			it('should request the correct url', () => {
				expect(fetch).toHaveBeenCalledWith(
					'https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=6&solrTerm=Portugal'
				);
			});

			it('should render a combobox', async () => {
				responseMock.results.docs.forEach(({ bookingId }) => {
					expect(screen.queryByTestId(bookingId)).toBeInTheDocument();
				});
			});

			describe('and user selects a combobox suggestion', () => {
				it('should update the input value', () => {
					const option = screen.queryByText('Lisbon Airport (LIS)');

					fireEvent.click(option);

					expect(input.value).toBe(
						'Lisbon Airport (LIS) , Lisbon, Lisbon Region, Portugal'
					);
				});
			});
		});
	});
});
