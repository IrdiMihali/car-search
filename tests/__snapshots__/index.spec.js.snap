// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given \`Home\` should render correctly 1`] = `
.c0 {
  background-size: cover;
  background-position: inherit;
  display: block;
}

.c0 main {
  display: block;
  height: inherit;
  background: #1879ca;
}

.c1 h1 {
  color: #fff;
  margin: 0;
  padding: 16px;
  text-align: left;
  line-height: 35px;
}

.c2 {
  color: #fff;
  padding: 4px 16px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1em;
  gap: 10px;
  list-style: none;
}

.c2 li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 3px;
  font-size: 0.9em;
}

.c2 li svg {
  fill: #fff;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  max-width: 1142px;
  padding: 0 16px;
  position: relative;
}

.c3 > div:first-of-type {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: var(--input-height);
  position: relative;
  width: 100%;
}

.c3 > div:first-of-type svg {
  fill: #000;
  height: 16px;
  left: 16px;
  pointer-events: none;
  position: absolute;
  right: auto;
  width: 16px;
  z-index: 1;
}

.c4 {
  border-radius: 10px 10px 0px 0px;
  border: 4px solid var(--main-border-color);
  box-sizing: border-box;
  font-size: 1em;
  font-weight: 400;
  height: 100%;
  left: 0;
  padding: 25px 40px 24px 40px;
  top: 0;
  width: 100%;
}

.c4:focus {
  outline-width: 0;
}

.c4:focus::-webkit-input-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c4:focus::-moz-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c4:focus:-ms-input-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c4:focus::placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c4:hover {
  outline-width: 0;
}

.c4::-webkit-search-cancel-button,
.c4::-webkit-search-results-decoration {
  display: none;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  max-width: 1142px;
  padding: 0 16px;
  position: relative;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #068323 !important;
  border-bottom: 4px solid var(--main-border-color);
  border-left: 4px solid var(--main-border-color);
  border-radius: 0px 0px 10px 10px;
  border-right: 4px solid var(--main-border-color);
  border-top: none;
  box-sizing: border-box;
  color: #fff;
  cursor: pointer;
  font-size: 1.4em;
  font-weight: 700;
  height: var(--input-height);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: normal;
  margin: 0;
  min-height: 44px;
  overflow: visible;
  padding: calc(4px * 2) 16px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color ease-in-out 0.1s,color ease-in-out 0.1s;
  transition: background-color ease-in-out 0.1s,color ease-in-out 0.1s;
  vertical-align: middle;
}

.c5:active {
  background-color: #056b1d;
}

@media (min-width:1024px) {
  .c0 main {
    background: linear-gradient(#1879ca,rgba(24,121,202,0));
  }
}

@media (min-width:1024px) {
  .c0 {
    background-image: url('../background-large.jpg');
    background-size: cover;
    background-repeat: no-repeat;
    background-position: 50%;
  }
}

@media (min-height:550px) and (min-width:1024px) {
  .c0 {
    height: min(calc(100vh),596px);
  }
}

@media (min-width:1024px) {
  .c1 h1 {
    font-size: 40px;
    line-height: 52px;
    padding-bottom: calc(4px * 2);
    padding-top: calc(4px * 6);
  }
}

@media (min-width:1024px) {
  .c1 {
    box-sizing: border-box;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin-left: auto;
    margin-right: auto;
    max-width: 1142px;
  }
}

@media (min-width:1024px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:1024px) {
  .c3 {
    padding-top: 16px;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:1024px) {
  .c4 {
    border-radius: 10px 0px 0px 10px;
  }
}

@media (min-width:1024px) {
  .c6 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:1024px) {
  .c5 {
    border-bottom: 4px solid var(--main-border-color);
    border-left: none;
    border-radius: 0px 10px 10px 0px;
    border-right: 4px solid var(--main-border-color);
    border-top: 4px solid var(--main-border-color);
    display: block;
  }
}

<div>
  <div
    class="c0"
  >
    <main>
      <div
        class="c1"
      >
        <h1>
          Car Hire â€“ Search, Compare & Save
        </h1>
        <ul
          class="c2 advantages-list"
        >
          <li>
            <svg
              height="1em"
              viewBox="0 0 128 128"
              width="1em"
            >
              <path
                d="M56.33 100a4 4 0 0 1-2.82-1.16L20.68 66.12a4 4 0 1 1 5.64-5.65l29.57 29.46 45.42-60.33a4 4 0 1 1 6.38 4.8l-48.17 64a4 4 0 0 1-2.91 1.6z"
              />
            </svg>
            Free cancelation on most bookings
          </li>
          <li>
            <svg
              height="1em"
              viewBox="0 0 128 128"
              width="1em"
            >
              <path
                d="M56.33 100a4 4 0 0 1-2.82-1.16L20.68 66.12a4 4 0 1 1 5.64-5.65l29.57 29.46 45.42-60.33a4 4 0 1 1 6.38 4.8l-48.17 64a4 4 0 0 1-2.91 1.6z"
              />
            </svg>
            60,000+ locations
          </li>
          <li>
            <svg
              height="1em"
              viewBox="0 0 128 128"
              width="1em"
            >
              <path
                d="M56.33 100a4 4 0 0 1-2.82-1.16L20.68 66.12a4 4 0 1 1 5.64-5.65l29.57 29.46 45.42-60.33a4 4 0 1 1 6.38 4.8l-48.17 64a4 4 0 0 1-2.91 1.6z"
              />
            </svg>
            Customer support in 40+ languages
          </li>
        </ul>
        <form>
          <div
            class="c3"
          >
            <div
              aria-activedescendant="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              data-testid="results-combobox"
              role="combobox"
            >
              <svg
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
              >
                <path
                  d="M17.464 6.56a8.313 8.313 0 1 1-15.302 6.504A8.313 8.313 0 0 1 17.464 6.56zm1.38-.586C16.724.986 10.963-1.339 5.974.781.988 2.9-1.337 8.662.783 13.65c2.12 4.987 7.881 7.312 12.87 5.192 4.987-2.12 7.312-7.881 5.192-12.87zM15.691 16.75l7.029 7.03a.75.75 0 0 0 1.06-1.06l-7.029-7.03a.75.75 0 0 0-1.06 1.06z"
                />
              </svg>
              <input
                aria-autocomplete=""
                aria-controls="location-search-listbox"
                aria-label="Pick-up location"
                class="c4"
                data-testid="search-input"
                id="location-search"
                name="q"
                placeholder="Pick-up location"
                type="search"
                value=""
              />
            </div>
            <button
              class="c5"
              type="submit"
            >
              Search
            </button>
          </div>
          <div
            class="c6"
          />
        </form>
      </div>
    </main>
  </div>
</div>
`;
